File: objc/Project2/Project2Tests/ThrowingTests.m
//
//  ThrowingTests.m
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

#import <XCTest/XCTest.h>

typedef NS_ENUM(NSUInteger, GameError) {
    GameErrorNotPurchased, GameErrorNotInstalled, GameErrorParentalControlsDisallowed
};

@interface Game : NSObject
@property NSString *name;
- (void)play:(NSError * __autoreleasing _Nullable *)error;
@end

@implementation Game
- (Game *)initWithName:(NSString *)name {
    self = [super init];
    if (self) self.name = name;
    return self;
}

- (void)play:(NSError *__autoreleasing  _Nullable *)error {
    if ([self.name isEqualToString:@"BioBlitz"]) {
        *error = [[NSError alloc] initWithDomain:@"GameErrorNotPurchased" code:GameErrorNotPurchased userInfo:@{NSLocalizedDescriptionKey: @"GameErrorNotPurchased"}];
    } else if ([self.name isEqualToString:@"Blastazap"]) {
        *error = [[NSError alloc] initWithDomain:@"GameErrorNotInstalled" code:GameErrorNotInstalled userInfo:@{NSLocalizedDescriptionKey: @"GameErrorNotInstalled"}];
    } else if ([self.name isEqualToString:@"Dead Storm Rising"]) {
        *error = [[NSError alloc] initWithDomain:@"GameErrorParentalControlsDisallowed" code:GameErrorParentalControlsDisallowed userInfo:@{NSLocalizedDescriptionKey: @"GameErrorParentalControlsDisallowed"}];
    } else {
        NSLog(@"%@ is OK to play!", self.name);
    }
}
@end

@interface ThrowingTests : XCTestCase

@end

@implementation ThrowingTests
- (void)testPlayingBioBlitzThrows {
    Game *game = [[Game alloc] initWithName:@"BioBlitz"];
    
    NSError * __autoreleasing _Nullable error = nil;
    [game play:&error];
    
    if (error == nil) {
        XCTFail(@"BioBlitz has not been purchased");
    } else if (error.code == GameErrorNotPurchased) {
        // success!
    } else {
        XCTFail();
    }
}
@end

File: swift/Project2/Project2Tests/ThrowingTests.swift
//
//  ThrowingTests.swift
//  Project2Tests
//
//  Created by Jinwoo Kim on 2/24/21.
//

import XCTest
@testable import Project2

enum GameError: Error {
    case notPurchased, notInstalled, parentalControlsDisallowed
}

struct Game: LocalizedError {
    let name: String
    
    func play() throws {
        if name == "BioBlitz" {
            throw GameError.notPurchased
        } else if name == "Blastazap" {
            throw GameError.notInstalled
        } else if name == "Dead Storm Rising" {
            throw GameError.parentalControlsDisallowed
        } else {
            print("\(name) is OK to play!")
        }
    }
}

// 나오는 에러 문구를 더 깔끔하게 만들어준다.
extension LocalizedError {
    var errorDescription: String? {
        return "\(self)"
    }
}

class ThrowingTests: XCTestCase {
    func testPlayingBioBlitzThrows() {
        let game = Game(name: "BioBlitz")
        
        do {
            try game.play()
            XCTFail("BioBlitz has not been purchased.")
        } catch GameError.notPurchased {
            // success!
        } catch {
            XCTFail()
        }
    }
    
    // 위 method를 요약하면
    func testPlayingBlastazapThrows() {
        let game = Game(name: "Blastazap")
        XCTAssertThrowsError(try game.play())
    }
    
    // 위 method를 요약하면
    func testPlayingBlastazapThrows2() {
        let game = Game(name: "Blastazap")
        XCTAssertThrowsError(try game.play()) { error in
            // Error Handling
            XCTAssertEqual(error as? GameError, GameError.notInstalled)
        }
    }
    
    // 에러가 없을 경우에만 Pass
    func testPlayingExplodingMonkeysDoesntThrow() {
        let game = Game(name: "Exploding Monkeys")
        XCTAssertNoThrow(try game.play())
    }
    
    // 에러가 없을 경우에만 Pass
    func testDeadStornRisingThrows() throws {
        let game = Game(name: "Dead Storm Rising")
        try game.play()
    }
    
    func testCrashyPlaneDoesntThrow() throws {
        let game = Game(name: "CrashyPlane")
        try game.play()
    }
}
